chain/client/src/stateless_validation/partial_witness_tracker.rs

```diff
@@ -0,0 +1,167 @@
+/// of the partial_witness so the LRU cache size need not be too large.
+const NUM_CHUNKS_IN_WITNESS_TRACKER_CACHE: usize = 200;
+
+/// Ratio of the number of data parts to total parts in the Reed Solomon encoding.
+/// The tradeoff here is having a higher ratio is better for handling missing parts and network errors
+/// but increases the size of the encoded state witness and the total network bandwidth requirements.
+const RATIO_DATA_PARTS: f32 = 0.8;
+
+/// Reed Solomon encoder for encoding state witness parts.
+/// We keep one wrapper for each length of chunk_validators to avoid re-creating the encoder.
+/// This is used by `PartialEncodedStateWitnessTracker`
+pub struct RsMap {
+    rs_map: HashMap<usize, ReedSolomonWrapper>,
+}
+
+impl RsMap {
+    pub fn new() -> Self {
+        Self { rs_map: HashMap::new() }
+    }
+
+    pub fn entry(&mut self, total_parts: usize) -> &ReedSolomonWrapper {
+        self.rs_map.entry(total_parts).or_insert_with(|| {
+            let data_parts = std::cmp::max((total_parts as f32 * RATIO_DATA_PARTS) as usize, 1);
+            ReedSolomonWrapper::new(data_parts, total_parts - data_parts)
+        })
+    }
+}
+
+struct CacheEntry {
+    pub is_decoded: bool,
+    pub data_parts_present: usize,
+    pub data_parts_required: usize,
+    pub parts: Vec<Option<Box<[u8]>>>,
+}
+
+impl CacheEntry {
+    pub fn new(total_parts: usize, data_parts: usize) -> Self {
+        Self {
+            is_decoded: false,
+            data_parts_present: 0,
+            data_parts_required: data_parts,
+            parts: vec![None; total_parts],
+        }
+    }
+
+    // Function to insert a part into the cache entry for the chunk hash. Additionally, it tries to
+    // decode and return the state witness if all parts are present.
+    pub fn insert_in_cache_entry(
+        &mut self,
+        partial_witness: PartialEncodedStateWitness,
+        rs: &ReedSolomonWrapper,
+    ) -> Option<EncodedChunkStateWitness> {
+        let chunk_hash = partial_witness.chunk_header().chunk_hash();
+        let (part_ord, part, encoded_length) = partial_witness.decompose();
+
+        // Check if the part is already present.
+        if self.parts[part_ord].is_some() {
+            tracing::warn!(
+                target: "stateless_validation",
+                ?chunk_hash,
+                part_ord = ?part_ord,
+                "Received duplicate or redundant partial state witness part."
+            );
+            return None;
+        }
+
+        // Check if we have already decoded the state witness.
+        if self.is_decoded {
+            return None;
+        }
+
+        // Insert the part into the cache entry.
+        self.parts[part_ord] = Some(part);
+        self.data_parts_present += 1;
+
+        // If we have enough parts, try to decode the state witness.
+        if self.data_parts_present < self.data_parts_required {
+            return None;
+        }
+        self.is_decoded = true;
```

Comment:
move this to Ok branch below? if we go into Err branch, I see that more parts can come and we may be able to decode later using these?

Reply:
Right, can do, but in all honesty, reed solomon erasure encoding isn't resistant to incorrect parts but only missing parts. In case we are not able to decode the witness, we probably would never be able to decode it.

Reply:
Done in a future PR https://github.com/near/nearcore/pull/11168/files#r1582677632